File: frontend/src/api/bundles.ts
import apiClient from './client'
import type {
  Bundle,
  BundleWithItems,
  BundleCreate,
  BundleUpdate,
  BundleItem,
  BundleItemCreate,
  PaginatedBundles,
  BundlePriceCalc,
} from '../types/product'

export const bundlesApi = {
  // List — Backend retourne PaginatedResponse: { items: [...], total, skip, limit }
  getBundles: async (params?: {
    page?: number
    page_size?: number
    featured?: boolean
    active_only?: boolean
  }): Promise<PaginatedBundles> => {
    const pageSize = params?.page_size || 20
    const page = params?.page || 1
    const backendParams: Record<string, unknown> = {
      skip: (page - 1) * pageSize,
      limit: pageSize,
    }
    if (params?.featured !== undefined) backendParams.featured = params.featured
    if (params?.active_only !== undefined) backendParams.active_only = params.active_only

    const { data } = await apiClient.get('/bundles', { params: backendParams })
    const items = Array.isArray(data.items) ? data.items : []
    const total = data.total ?? 0
    return {
      items,
      total,
      page,
      page_size: pageSize,
      total_pages: Math.ceil(total / pageSize) || 0,
    }
  },

  getBundle: async (id: number): Promise<BundleWithItems> => {
    const { data } = await apiClient.get(`/bundles/${id}`)
    return data.data || data
  },

  createBundle: async (bundle: BundleCreate): Promise<Bundle> => {
    const { data } = await apiClient.post('/bundles', bundle)
    return data.data || data
  },

  updateBundle: async (id: number, bundle: BundleUpdate): Promise<Bundle> => {
    const { data } = await apiClient.patch(`/bundles/${id}`, bundle)
    return data.data || data
  },

  deleteBundle: async (id: number): Promise<void> => {
    await apiClient.delete(`/bundles/${id}`)
  },

  // Items
  addItem: async (
    bundleId: number,
    item: BundleItemCreate
  ): Promise<BundleItem> => {
    const { data } = await apiClient.post(`/bundles/${bundleId}/items`, item)
    return data.data || data
  },

  updateItem: async (
    bundleId: number,
    itemId: number,
    item: { quantity: number }
  ): Promise<BundleItem> => {
    const { data } = await apiClient.patch(`/bundles/${bundleId}/items/${itemId}`, item)
    return data.data || data
  },

  removeItem: async (bundleId: number, itemId: number): Promise<void> => {
    await apiClient.delete(`/bundles/${bundleId}/items/${itemId}`)
  },

  calculatePrice: async (bundleId: number): Promise<BundlePriceCalc> => {
    const { data } = await apiClient.get(`/bundles/${bundleId}/calculate-price`)
    return data.data || data
  },
}

File: app/api/v1/endpoints/bundles.py
"""Endpoints CRUD pour les bundles (packs de produits)."""
import logging
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import joinedload
from app.core.database import get_async_db
from app.core.deps import get_current_user, require_scope, UserCompat
from app.core.permissions import Scope
from app.services.bundle import BundleService
from app.schemas.bundle import (
    BundleCreate,
    BundleUpdate,
    BundleResponse,
    BundleWithItems,
    BundleItemCreate,
    BundleItemUpdate,
    BundleItemResponse,
    BundlePriceResponse,
)
from app.schemas.common import PaginationParams, PaginatedResponse
from app.constants import ErrorMessages
from app.models.bundle import BundleItem as BundleItemModel


logger = logging.getLogger(__name__)

router = APIRouter(prefix="/bundles", tags=["Bundles"])


@router.get("", response_model=PaginatedResponse[BundleResponse])
async def list_bundles(
    pagination: PaginationParams = Depends(),
    featured: bool = Query(None, description="Filtrer bundles mis en avant"),
    active_only: bool = Query(True, description="Filtrer bundles actifs uniquement"),
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_READ)),
) -> PaginatedResponse[BundleResponse]:
    """Liste les bundles avec pagination."""
    service = BundleService(db)

    bundles, total = await service.list_bundles(
        tenant_id=current_user.tenant_id,
        skip=pagination.skip,
        limit=pagination.limit,
        featured=featured,
        include_inactive=not active_only,
    )

    return PaginatedResponse(
        items=[BundleResponse.model_validate(b) for b in bundles],
        total=total,
        skip=pagination.skip,
        limit=pagination.limit,
    )


@router.get("/{bundle_id}", response_model=BundleWithItems)
async def get_bundle(
    bundle_id: int,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_READ)),
) -> BundleWithItems:
    """Recupere les details d'un bundle avec ses items."""
    service = BundleService(db)
    bundle = await service.get_bundle(bundle_id, current_user.tenant_id)
    return BundleWithItems.model_validate(bundle)


@router.post("", response_model=BundleResponse, status_code=status.HTTP_201_CREATED)
async def create_bundle(
    data: BundleCreate,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_WRITE)),
) -> BundleResponse:
    """Cree un nouveau bundle (admin uniquement)."""
    service = BundleService(db)
    try:
        bundle = await service.create_bundle(data, current_user.tenant_id)
        await db.commit()
        await db.refresh(bundle)
        return BundleResponse.model_validate(bundle)
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Unexpected error in create_bundle")
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating bundle: {str(e)}",
        )


@router.patch("/{bundle_id}", response_model=BundleResponse)
async def update_bundle(
    bundle_id: int,
    data: BundleUpdate,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_WRITE)),
) -> BundleResponse:
    """Met a jour un bundle (admin uniquement)."""
    service = BundleService(db)
    try:
        bundle = await service.update_bundle(bundle_id, data, current_user.tenant_id)
        await db.commit()
        await db.refresh(bundle)
        return BundleResponse.model_validate(bundle)
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Unexpected error in update_bundle")
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error updating bundle: {str(e)}",
        )


@router.delete("/{bundle_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_bundle(
    bundle_id: int,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_DELETE)),
) -> None:
    """Soft delete un bundle (admin uniquement)."""
    service = BundleService(db)
    try:
        await service.delete_bundle(bundle_id, current_user.tenant_id)
        await db.commit()
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Unexpected error in delete_bundle")
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error deleting bundle: {str(e)}",
        )


@router.post(
    "/{bundle_id}/items",
    response_model=BundleItemResponse,
    status_code=status.HTTP_201_CREATED
)
async def add_bundle_item(
    bundle_id: int,
    data: BundleItemCreate,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_WRITE)),
) -> BundleItemResponse:
    """Ajoute un produit au bundle (admin uniquement)."""
    service = BundleService(db)
    try:
        item = await service.add_item(bundle_id, data, current_user.tenant_id)
        await db.commit()
        await db.refresh(item)
        reloaded = await db.get(BundleItemModel, item.id, options=[joinedload(BundleItemModel.product)])
        return BundleItemResponse.model_validate(reloaded)
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Unexpected error in add_bundle_item")
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error adding item: {str(e)}",
        )


@router.patch("/{bundle_id}/items/{item_id}", response_model=BundleItemResponse)
async def update_bundle_item(
    bundle_id: int,
    item_id: int,
    data: BundleItemUpdate,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_WRITE)),
) -> BundleItemResponse:
    """Met a jour un item du bundle (admin uniquement)."""
    service = BundleService(db)
    try:
        item = await service.update_item(bundle_id, item_id, data, current_user.tenant_id)
        await db.commit()
        await db.refresh(item)
        reloaded = await db.get(BundleItemModel, item.id, options=[joinedload(BundleItemModel.product)])
        return BundleItemResponse.model_validate(reloaded)
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Unexpected error in update_bundle_item")
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error updating item: {str(e)}",
        )


@router.delete("/{bundle_id}/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_bundle_item(
    bundle_id: int,
    item_id: int,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_DELETE)),
) -> None:
    """Supprime un item du bundle (admin uniquement)."""
    service = BundleService(db)
    try:
        await service.remove_item(bundle_id, item_id, current_user.tenant_id)
        await db.commit()
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Unexpected error in remove_bundle_item")
        await db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error removing item: {str(e)}",
        )


@router.get("/{bundle_id}/calculate-price", response_model=BundlePriceResponse)
async def calculate_bundle_price(
    bundle_id: int,
    db: AsyncSession = Depends(get_async_db),
    current_user: UserCompat = Depends(require_scope(Scope.BUNDLES_READ)),
) -> BundlePriceResponse:
    """Calcule le prix individuel vs prix bundle."""
    service = BundleService(db)
    result = await service.calculate_price(bundle_id, current_user.tenant_id)
    return BundlePriceResponse(**result)
